### PR TITLE
Rule S1151: "switch case" clauses should not have too many lines of code

### LINKED ISSUE #1170: Rule S1151: "switch case" clauses should not have too many lines of code
Implements [RSPEC-1151](https://jira.sonarsource.com/browse/RSPEC-1151)

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #1170

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                                Console.WriteLine();
+                                Console.WriteLine();
+                            }
+                        }
+                    }
+                case 5: // Noncompliant
+                    var x = Foo(42);
+                    break;
+
+                    string Foo(int value)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs]: Not sure what to do with local functions. Shall we ignore them as we recommend to use a method? Shall I skip the function declaration (e.g. `string Foo(int value)`) from the count?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SwitchSectionShouldNotHaveTooManyStatements.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        protected override void Initialize(ParameterLoadingAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var switchSection = (SwitchSectionSyntax)c.Node;
+
+                    if (switchSection.IsMissing ||
+                        switchSection.Labels.Count <= 0)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SwitchSectionShouldNotHaveTooManyStatements.cs]: How could this be less than 0?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                case 1: // Noncompliant {{Reduce this 'switch/case' number of lines from 6 to at most 5, for example by extracting code into a method.}}
+//              ^^^^^^^
+                    Console.WriteLine("1");
+                    Console.WriteLine("2");
+                    Console.WriteLine("3");
+                    Console.WriteLine("4");
+                    Console.WriteLine("5");
+                    break;
+                case 2:
+                    {
```
Comment (valhristov) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs]: Do we count blocks as one line? If yes, this test should be negative.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs (hunk 1 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs]: The rule is about readability - blocks with too many lines are difficult to read, we should not ignore locals

**Comment by valhristov** [comment]
But I guess you need to run the QA before merging :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SwitchSectionShouldNotHaveTooManyStatements.cs (hunk 2 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SwitchSectionShouldNotHaveTooManyStatements.cs]: This shouldn't be... Just playing defensive as usual.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs (hunk 3 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_DefaultValue.cs]: The statements collection here will return only one statement, the block but as you said this is about readability so I thought I should expand blocks hence this count. Shall I change this?

**Pull request merged**